File: 0070-climbing-stairs/0070-climbing-stairs.cpp
class Solution {
    private:
    int f(int stair,int n,vector<int>& dp){
        if(stair > n) return 0;
        if(stair == n) return 1;
        if(dp[stair]!=-1) return dp[stair];
        int first = f(stair+1,n,dp);
        int second = f(stair+2,n,dp);
        return dp[stair] = first+second;
    }
public:
    int climbStairs(int n) {
        vector<int> dp(n+1,-1);
        return f(0,n,dp);
        
    }
};

File: 0070-climbing-stairs/0070-climbing-stairs.py
class Solution:
    def tot_ways_to_climb(self, n:int, dp: List[int]) -> int:
        if n == 0:
            return 1
        if n < 0:
            return 0
        if dp[n] != -1:
            return dp[n]
        take_one_step = self.tot_ways_to_climb(n-1, dp)
        take_two_steps = self.tot_ways_to_climb(n-2, dp)
        ans =  take_one_step + take_two_steps
        dp[n] = ans
        return ans
    def climbStairs(self, n: int) -> int:
        dp = [-1]*(n+1)
        return self.tot_ways_to_climb(n, dp)
        
        
        